File: app_pojav_zh/src/main/java/com/movtery/pojavzh/ui/fragment/ModsFragment.java
package com.movtery.pojavzh.ui.fragment;

import static com.movtery.pojavzh.utils.file.FileTools.copyFileInBackground;
import static net.kdt.pojavlaunch.Tools.runOnUiThread;

import android.os.Bundle;
import android.view.View;
import android.widget.CheckBox;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;

import com.daimajia.androidanimations.library.Techniques;
import com.movtery.pojavzh.ui.dialog.FilesDialog;
import com.movtery.pojavzh.ui.subassembly.filelist.FileIcon;
import com.movtery.pojavzh.ui.subassembly.filelist.FileRecyclerAdapter;
import com.movtery.pojavzh.ui.subassembly.filelist.FileRecyclerView;
import com.movtery.pojavzh.ui.subassembly.filelist.FileSelectedListener;
import com.movtery.pojavzh.ui.subassembly.view.SearchView;
import com.movtery.pojavzh.utils.anim.AnimUtils;
import com.movtery.pojavzh.utils.ZHTools;
import com.movtery.pojavzh.utils.anim.OnSlideOutListener;
import com.movtery.pojavzh.utils.anim.ViewAnimUtils;
import com.movtery.pojavzh.utils.file.FileTools;
import com.movtery.pojavzh.utils.file.OperationFile;
import com.movtery.pojavzh.utils.file.PasteFile;

import net.kdt.pojavlaunch.PojavApplication;
import net.kdt.pojavlaunch.R;
import net.kdt.pojavlaunch.contracts.OpenDocumentWithExtension;
import net.kdt.pojavlaunch.fragments.SearchModFragment;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class ModsFragment extends FragmentWithAnim {
    public static final String TAG = "ModsFragment";
    public static final String BUNDLE_ROOT_PATH = "root_path";
    public static final String jarFileSuffix = ".jar";
    public static final String disableJarFileSuffix = ".jar.disabled";
    private ActivityResultLauncher<Object> openDocumentLauncher;
    private View mModsLayout, mOperateLayout, mShadowView, mOperateView;
    private ImageButton mReturnButton, mAddModButton, mPasteButton, mDownloadButton, mSearchSummonButton, mRefreshButton;
    private TextView mNothingTip;
    private SearchView mSearchView;
    private CheckBox mMultiSelectCheck, mSelectAllCheck;
    private FileRecyclerView mFileRecyclerView;
    private String mRootPath;

    public ModsFragment() {
        super(R.layout.fragment_mods);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        openDocumentLauncher = registerForActivityResult(
                new OpenDocumentWithExtension("jar"),
                result -> {
                    if (result != null) {
                        Toast.makeText(requireContext(), getString(R.string.tasks_ongoing), Toast.LENGTH_SHORT).show();

                        PojavApplication.sExecutorService.execute(() -> {
                            copyFileInBackground(requireContext(), result, mRootPath);

                            runOnUiThread(() -> {
                                Toast.makeText(requireContext(), getString(R.string.zh_profile_mods_added_mod), Toast.LENGTH_SHORT).show();
                                mFileRecyclerView.refreshPath();
                            });
                        });
                    }
                }
        );
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        bindViews(view);
        parseBundle();

        mFileRecyclerView.setShowFiles(true);
        mFileRecyclerView.setShowFolders(false);
        mFileRecyclerView.lockAndListAt(new File(mRootPath), new File(mRootPath));

        mFileRecyclerView.setFileSelectedListener(new FileSelectedListener() {
            @Override
            public void onFileSelected(File file, String path) {
                showDialog(file);
            }

            @Override
            public void onItemLongClick(File file, String path) {
                if (file.isDirectory()) {
                    showDialog(file);
                }
            }
        });
        mFileRecyclerView.setOnMultiSelectListener(itemBeans -> {
            if (!itemBeans.isEmpty()) {
                PojavApplication.sExecutorService.execute(() -> {
                });
                //取出全部文件
                List<File> selectedFiles = new ArrayList<>();
                itemBeans.forEach(value -> {
                    File file = value.getFile();
                    if (file != null) {
                        selectedFiles.add(file);
                    }
                });
                FilesDialog.FilesButton filesButton = new FilesDialog.FilesButton();
                filesButton.setButtonVisibility(true, true, false, false, true, true);
                filesButton.setDialogText(getString(R.string.zh_file_multi_select_mode_title),
                        getString(R.string.zh_file_multi_select_mode_message, itemBeans.size()),
                        getString(R.string.zh_profile_mods_disable_or_enable));
                runOnUiThread(() -> {
                    FilesDialog filesDialog = new FilesDialog(requireContext(), filesButton, () -> runOnUiThread(() -> {
                        closeMultiSelect();
                        mFileRecyclerView.refreshPath();
                    }), selectedFiles);
                    filesDialog.setCopyButtonClick(() -> mPasteButton.setVisibility(View.VISIBLE));
                    filesDialog.setMoreButtonClick(() -> new OperationFile(requireContext(), () -> runOnUiThread(() -> {
                        closeMultiSelect();
                        mFileRecyclerView.refreshPath();
                    }), file -> {
                        if (file != null && file.exists()) {
                            String fileName = file.getName();
                            if (fileName.endsWith(jarFileSuffix)) {
                                disableMod(file);
                            } else if (fileName.endsWith(disableJarFileSuffix)) {
                                enableMod(file);
                            }
                        }
                    }).operationFile(selectedFiles));
                    filesDialog.show();
                });
            }
        });
        mFileRecyclerView.setRefreshListener(() -> {
            int itemCount = mFileRecyclerView.getItemCount();
            boolean show = itemCount == 0;
            AnimUtils.setVisibilityAnim(mNothingTip, show);
        });
        FileRecyclerAdapter adapter = mFileRecyclerView.getAdapter();
        mMultiSelectCheck.setOnCheckedChangeListener((buttonView, isChecked) -> {
            mSelectAllCheck.setChecked(false);
            mSelectAllCheck.setVisibility(isChecked ? View.VISIBLE : View.GONE);
            adapter.setMultiSelectMode(isChecked);
        });
        mSelectAllCheck.setOnCheckedChangeListener((buttonView, isChecked) -> adapter.selectAllFiles(isChecked));

        mReturnButton.setOnClickListener(v -> {
            closeMultiSelect();
            ZHTools.onBackPressed(requireActivity());
        });
        mAddModButton.setOnClickListener(v -> {
            closeMultiSelect();
            String suffix = ".jar";
            Toast.makeText(requireActivity(), String.format(getString(R.string.zh_file_add_file_tip), suffix), Toast.LENGTH_SHORT).show();
            openDocumentLauncher.launch(suffix);
        });
        mPasteButton.setOnClickListener(v -> PasteFile.getInstance().pasteFiles(requireActivity(), mFileRecyclerView.getFullPath(), this::getFileSuffix, () -> runOnUiThread(() -> {
            closeMultiSelect();
            mPasteButton.setVisibility(View.GONE);
            mFileRecyclerView.refreshPath();
        })));
        mDownloadButton.setOnClickListener(v -> {
            closeMultiSelect();
            Bundle bundle = new Bundle();
            bundle.putBoolean(SearchModFragment.BUNDLE_SEARCH_MODPACK, false);
            bundle.putString(SearchModFragment.BUNDLE_MOD_PATH, mRootPath);
            ZHTools.swapFragmentWithAnim(this, SearchModFragment.class, SearchModFragment.TAG, bundle);
        });
        mSearchSummonButton.setOnClickListener(v -> {
            closeMultiSelect();
            mSearchView.setVisibility();
        });
        mRefreshButton.setOnClickListener(v -> {
            closeMultiSelect();
            mFileRecyclerView.refreshPath();
        });

        ViewAnimUtils.slideInAnim(this);
    }

    private void closeMultiSelect() {
        //点击其它控件时关闭多选模式
        mMultiSelectCheck.setChecked(false);
        mSelectAllCheck.setVisibility(View.GONE);
    }

    private void showDialog(File file) {
        String fileName = file.getName();

        FilesDialog.FilesButton filesButton = new FilesDialog.FilesButton();
        filesButton.setButtonVisibility(true, true, !file.isDirectory(), true, true, (fileName.endsWith(jarFileSuffix) || fileName.endsWith(disableJarFileSuffix)));
        if (file.isDirectory()) {
            filesButton.setMessageText(getString(R.string.zh_file_folder_message));
        } else {
            filesButton.setMessageText(getString(R.string.zh_file_message));
        }
        if (fileName.endsWith(jarFileSuffix))
            filesButton.setMoreButtonText(getString(R.string.zh_profile_mods_disable));
        else if (fileName.endsWith(disableJarFileSuffix))
            filesButton.setMoreButtonText(getString(R.string.zh_profile_mods_enable));

        FilesDialog filesDialog = new FilesDialog(requireContext(), filesButton, () -> runOnUiThread(() -> mFileRecyclerView.refreshPath()), file);

        filesDialog.setCopyButtonClick(() -> mPasteButton.setVisibility(View.VISIBLE));

        //检测后缀名，以设置正确的按钮
        if (fileName.endsWith(jarFileSuffix)) {
            filesDialog.setFileSuffix(jarFileSuffix);
            filesDialog.setMoreButtonClick(() -> {
                disableMod(file);
                mFileRecyclerView.refreshPath();
                filesDialog.dismiss();
            });
        } else if (fileName.endsWith(disableJarFileSuffix)) {
            filesDialog.setFileSuffix(disableJarFileSuffix);
            filesDialog.setMoreButtonClick(() -> {
                enableMod(file);
                mFileRecyclerView.refreshPath();
                filesDialog.dismiss();
            });
        }

        filesDialog.show();
    }

    private void disableMod(File file) {
        String fileName = file.getName();
        String fileParent = file.getParent();
        File newFile = new File(fileParent, fileName + ".disabled");
        FileTools.renameFile(file, newFile);
    }

    private void enableMod(File file) {
        String fileName = file.getName();
        String fileParent = file.getParent();
        String newFileName = fileName.substring(0, fileName.lastIndexOf(disableJarFileSuffix));
        if (!fileName.endsWith(jarFileSuffix))
            newFileName += jarFileSuffix; //如果没有.jar结尾，那么默认加上.jar后缀

        File newFile = new File(fileParent, newFileName);
        FileTools.renameFile(file, newFile);
    }

    private String getFileSuffix(File file) {
        String name = file.getName();
        if (name.endsWith(disableJarFileSuffix)) {
            return disableJarFileSuffix;
        } else if (name.endsWith(jarFileSuffix)) {
            return jarFileSuffix;
        } else {
            int dotIndex = file.getName().lastIndexOf('.');
            return dotIndex == -1 ? "" : file.getName().substring(dotIndex);
        }
    }

    private void parseBundle() {
        Bundle bundle = getArguments();
        if (bundle == null) return;
        mRootPath = bundle.getString(BUNDLE_ROOT_PATH, mRootPath);
    }

    private void bindViews(@NonNull View view) {
        mModsLayout = view.findViewById(R.id.mods_layout);
        mOperateLayout = view.findViewById(R.id.operate_layout);
        mShadowView = view.findViewById(R.id.shadowView);

        mOperateView = view.findViewById(R.id.operate_view);

        mReturnButton = view.findViewById(R.id.zh_return_button);
        mAddModButton = view.findViewById(R.id.zh_add_file_button);
        mPasteButton = view.findViewById(R.id.zh_paste_button);
        mDownloadButton = view.findViewById(R.id.zh_create_folder_button);
        mSearchSummonButton = view.findViewById(R.id.zh_search_button);
        mRefreshButton = view.findViewById(R.id.zh_refresh_button);
        mNothingTip = view.findViewById(R.id.zh_mods_nothing);

        mAddModButton.setContentDescription(getString(R.string.zh_profile_mods_add_mod));
        mDownloadButton.setContentDescription(getString(R.string.zh_profile_mods_download_mod));
        mDownloadButton.setImageDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.ic_download));

        mMultiSelectCheck = view.findViewById(R.id.zh_mods_multi_select_files);
        mSelectAllCheck = view.findViewById(R.id.zh_mods_select_all);
        mFileRecyclerView = view.findViewById(R.id.zh_mods);

        mSearchView = new SearchView(view, view.findViewById(R.id.zh_search_view));
        mSearchView.setSearchListener(mFileRecyclerView::searchFiles);
        mSearchView.setShowSearchResultsListener(mFileRecyclerView::setShowSearchResultsOnly);
        mFileRecyclerView.setFileIcon(FileIcon.MOD);

        mPasteButton.setVisibility(PasteFile.getInstance().getPasteType() != null ? View.VISIBLE : View.GONE);

        ZHTools.setTooltipText(mReturnButton, mReturnButton.getContentDescription());
        ZHTools.setTooltipText(mAddModButton, mAddModButton.getContentDescription());
        ZHTools.setTooltipText(mPasteButton, mPasteButton.getContentDescription());
        ZHTools.setTooltipText(mDownloadButton, mDownloadButton.getContentDescription());
        ZHTools.setTooltipText(mSearchSummonButton, mSearchSummonButton.getContentDescription());
        ZHTools.setTooltipText(mRefreshButton, mRefreshButton.getContentDescription());
    }

    @Override
    public void slideIn() {
        ViewAnimUtils.setViewAnim(mModsLayout, Techniques.BounceInDown);
        ViewAnimUtils.setViewAnim(mOperateLayout, Techniques.BounceInLeft);
        ViewAnimUtils.setViewAnim(mShadowView, Techniques.BounceInLeft);
        ViewAnimUtils.setViewAnim(mOperateView, Techniques.FadeInLeft);
    }

    @Override
    public void slideOut(@NonNull OnSlideOutListener listener) {
        ViewAnimUtils.setViewAnim(mModsLayout, Techniques.FadeOutUp);
        ViewAnimUtils.setViewAnim(mOperateLayout, Techniques.FadeOutRight);
        ViewAnimUtils.setViewAnim(mShadowView, Techniques.FadeOutRight);
        super.slideOut(listener);
    }
}

File: app_pojav_zh/src/main/java/com/movtery/pojavzh/ui/fragment/ModsFragment.kt
package com.movtery.pojavzh.ui.fragment

import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CompoundButton
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.core.content.ContextCompat
import com.getkeepsafe.taptargetview.TapTargetSequence
import com.movtery.anim.AnimPlayer
import com.movtery.anim.animations.Animations
import com.movtery.pojavzh.feature.mod.ModUtils
import com.movtery.pojavzh.ui.dialog.FilesDialog
import com.movtery.pojavzh.ui.dialog.FilesDialog.FilesButton
import com.movtery.pojavzh.ui.dialog.TipDialog
import com.movtery.pojavzh.ui.subassembly.filelist.FileIcon
import com.movtery.pojavzh.ui.subassembly.filelist.FileItemBean
import com.movtery.pojavzh.ui.subassembly.filelist.FileSelectedListener
import com.movtery.pojavzh.ui.subassembly.view.SearchViewWrapper
import com.movtery.pojavzh.utils.ZHTools
import com.movtery.pojavzh.utils.anim.AnimUtils.Companion.setVisibilityAnim
import com.movtery.pojavzh.utils.file.FileCopyHandler
import com.movtery.pojavzh.utils.file.FileTools.Companion.copyFileInBackground
import com.movtery.pojavzh.feature.mod.ModToggleHandler
import com.movtery.pojavzh.utils.NewbieGuideUtils
import com.movtery.pojavzh.utils.file.PasteFile
import net.kdt.pojavlaunch.PojavApplication
import net.kdt.pojavlaunch.R
import net.kdt.pojavlaunch.Tools
import net.kdt.pojavlaunch.contracts.OpenDocumentWithExtension
import net.kdt.pojavlaunch.databinding.FragmentModsBinding
import net.kdt.pojavlaunch.fragments.SearchModFragment
import java.io.File
import java.util.function.Consumer

class ModsFragment : FragmentWithAnim(R.layout.fragment_mods) {
    companion object {
        const val TAG: String = "ModsFragment"
        const val BUNDLE_ROOT_PATH: String = "root_path"
    }

    private lateinit var binding: FragmentModsBinding
    private lateinit var mSearchViewWrapper: SearchViewWrapper
    private var openDocumentLauncher: ActivityResultLauncher<Any>? = null
    private var mRootPath: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        openDocumentLauncher = registerForActivityResult(OpenDocumentWithExtension("jar")) { result: Uri? ->
            result?.let{
                Toast.makeText(requireContext(), getString(R.string.tasks_ongoing), Toast.LENGTH_SHORT).show()

                PojavApplication.sExecutorService.execute {
                    copyFileInBackground(requireContext(), result, mRootPath)
                    Tools.runOnUiThread { Toast.makeText(requireContext(), getString(R.string.profile_mods_added_mod), Toast.LENGTH_SHORT).show()
                        binding.fileRecyclerView.refreshPath()
                    }
                }
            }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentModsBinding.inflate(layoutInflater)
        mSearchViewWrapper = SearchViewWrapper(this)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        initViews()
        parseBundle()

        binding.apply {
            fileRecyclerView.apply {
                setShowFiles(true)
                setShowFolders(false)

                setFileSelectedListener(object : FileSelectedListener() {
                    override fun onFileSelected(file: File?, path: String?) {
                        file?.let {
                            if (it.isFile) {
                                val fileName = it.name

                                val filesButton = FilesButton()
                                filesButton.setButtonVisibility(true, true, true, true, true,
                                    (fileName.endsWith(ModUtils.JAR_FILE_SUFFIX) || fileName.endsWith(ModUtils.DISABLE_JAR_FILE_SUFFIX)))
                                filesButton.setMessageText(if (it.isDirectory) getString(R.string.file_folder_message) else getString(R.string.file_message))

                                if (fileName.endsWith(ModUtils.JAR_FILE_SUFFIX)) filesButton.setMoreButtonText(getString(R.string.profile_mods_disable))
                                else if (fileName.endsWith(ModUtils.DISABLE_JAR_FILE_SUFFIX)) filesButton.setMoreButtonText(getString(R.string.profile_mods_enable))

                                val filesDialog = FilesDialog(requireContext(), filesButton,
                                    { Tools.runOnUiThread { refreshPath() } },
                                    fullPath, it
                                )

                                filesDialog.setCopyButtonClick { visibility = View.VISIBLE }

                                //检测后缀名，以设置正确的按钮
                                if (fileName.endsWith(ModUtils.JAR_FILE_SUFFIX)) {
                                    filesDialog.setFileSuffix(ModUtils.JAR_FILE_SUFFIX)
                                    filesDialog.setMoreButtonClick {
                                        ModUtils.disableMod(it)
                                        refreshPath()
                                        filesDialog.dismiss()
                                    }
                                } else if (fileName.endsWith(ModUtils.DISABLE_JAR_FILE_SUFFIX)) {
                                    filesDialog.setFileSuffix(ModUtils.DISABLE_JAR_FILE_SUFFIX)
                                    filesDialog.setMoreButtonClick {
                                        ModUtils.enableMod(it)
                                        refreshPath()
                                        filesDialog.dismiss()
                                    }
                                }

                                filesDialog.show()
                            }
                        }
                    }

                    override fun onItemLongClick(file: File?, path: String?) {
                    }
                })

                setOnMultiSelectListener { itemBeans: List<FileItemBean> ->
                    if (itemBeans.isNotEmpty()) {
                        PojavApplication.sExecutorService.execute {}
                        //取出全部文件
                        val selectedFiles: MutableList<File> = ArrayList()
                        itemBeans.forEach(Consumer { value: FileItemBean ->
                            val file = value.file
                            file?.apply { selectedFiles.add(this) }
                        })
                        val filesButton = FilesButton()
                        filesButton.setButtonVisibility(true, true, false, false, true, true)
                        filesButton.setDialogText(
                            getString(R.string.file_multi_select_mode_title),
                            getString(R.string.file_multi_select_mode_message, itemBeans.size),
                            getString(R.string.profile_mods_disable_or_enable)
                        )
                        Tools.runOnUiThread {
                            val filesDialog = FilesDialog(requireContext(), filesButton, {
                                Tools.runOnUiThread {
                                    closeMultiSelect()
                                    refreshPath()
                                }
                            }, fullPath, selectedFiles)
                            filesDialog.setCopyButtonClick { operateView.pasteButton.visibility = View.VISIBLE }
                            filesDialog.setMoreButtonClick {
                                ModToggleHandler(requireContext(), selectedFiles) {
                                    Tools.runOnUiThread {
                                        closeMultiSelect()
                                        refreshPath()
                                    }
                                }.start()
                            }
                            filesDialog.show()
                        }
                    }
                }

                setRefreshListener {
                    val show = itemCount == 0
                    setVisibilityAnim(nothingLayout, show)
                }
            }

            multiSelectFiles.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
                selectAll.apply {
                    this.isChecked = false
                    visibility = if (isChecked) View.VISIBLE else View.GONE
                }
                fileRecyclerView.adapter.setMultiSelectMode(isChecked)
                mSearchViewWrapper.let { if (mSearchViewWrapper.isVisible()) mSearchViewWrapper.setVisibility(!isChecked) }
            }

            operateView.apply {
                selectAll.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
                    fileRecyclerView.adapter.selectAllFiles(isChecked)
                }

                returnButton.setOnClickListener {
                    closeMultiSelect()
                    ZHTools.onBackPressed(requireActivity())
                }

                addFileButton.setOnClickListener {
                    closeMultiSelect()
                    val suffix = ".jar"
                    Toast.makeText(
                        requireActivity(),
                        String.format(getString(R.string.file_add_file_tip), suffix),
                        Toast.LENGTH_SHORT
                    ).show()
                    openDocumentLauncher?.launch(suffix)
                }

                pasteButton.setOnClickListener {
                    PasteFile.getInstance().pasteFiles(
                        requireActivity(),
                        fileRecyclerView.fullPath,
                        object : FileCopyHandler.FileExtensionGetter {
                            override fun onGet(file: File?): String? {
                                return file?.let { it1 -> getFileSuffix(it1) }
                            }
                        }
                    ) {
                        Tools.runOnUiThread {
                            closeMultiSelect()
                            pasteButton.visibility = View.GONE
                            fileRecyclerView.refreshPath()
                        }
                    }
                }

                createFolderButton.setOnClickListener { goDownloadMod() }

                downloadOptifine.setOnClickListener {
                    TipDialog.Builder(requireContext())
                        .setMessage(R.string.profile_manager_download_optifine_message)
                        .setConfirmClickListener {
                            val bundle = Bundle()
                            bundle.putBoolean(DownloadOptiFineFragment.BUNDLE_DOWNLOAD_MOD, true)
                            ZHTools.swapFragmentWithAnim(
                                this@ModsFragment,
                                DownloadOptiFineFragment::class.java,
                                DownloadOptiFineFragment.TAG,
                                bundle
                            )
                        }.buildDialog()
                }

                searchButton.setOnClickListener {
                    closeMultiSelect()
                    mSearchViewWrapper.setVisibility()
                }

                refreshButton.setOnClickListener {
                    closeMultiSelect()
                    fileRecyclerView.refreshPath()
                }
            }

            goDownloadText.setOnClickListener{ goDownloadMod() }

            fileRecyclerView.lockAndListAt(mRootPath?.let { File(it) }, mRootPath?.let { File(it) })
        }

        startNewbieGuide()
    }

    private fun startNewbieGuide() {
        if (NewbieGuideUtils.showOnlyOne(TAG)) return
        binding.operateView.apply {
            val fragmentActivity = requireActivity()
            TapTargetSequence(fragmentActivity)
                .targets(
                    NewbieGuideUtils.getSimpleTarget(fragmentActivity, refreshButton, getString(R.string.generic_refresh), getString(R.string.newbie_guide_general_refresh)),
                    NewbieGuideUtils.getSimpleTarget(fragmentActivity, searchButton, getString(R.string.generic_search), getString(R.string.newbie_guide_mod_search)),
                    NewbieGuideUtils.getSimpleTarget(fragmentActivity, addFileButton, getString(R.string.profile_mods_add_mod), getString(R.string.newbie_guide_mod_import)),
                    NewbieGuideUtils.getSimpleTarget(fragmentActivity, createFolderButton, getString(R.string.profile_mods_download_mod), getString(R.string.newbie_guide_mod_download)),
                    NewbieGuideUtils.getSimpleTarget(fragmentActivity, returnButton, getString(R.string.generic_close), getString(R.string.newbie_guide_general_close)))
                .start()
        }
    }

    private fun closeMultiSelect() {
        //点击其它控件时关闭多选模式
        binding.apply {
            multiSelectFiles.isChecked = false
            selectAll.visibility = View.GONE
        }
    }

    private fun getFileSuffix(file: File): String {
        val name = file.name
        if (name.endsWith(ModUtils.DISABLE_JAR_FILE_SUFFIX)) {
            return ModUtils.DISABLE_JAR_FILE_SUFFIX
        } else if (name.endsWith(ModUtils.JAR_FILE_SUFFIX)) {
            return ModUtils.JAR_FILE_SUFFIX
        } else {
            val dotIndex = file.name.lastIndexOf('.')
            return if (dotIndex == -1) "" else file.name.substring(dotIndex)
        }
    }

    private fun goDownloadMod() {
        closeMultiSelect()
        val bundle = Bundle()
        bundle.putBoolean(SearchModFragment.BUNDLE_SEARCH_MODPACK, false)
        bundle.putString(SearchModFragment.BUNDLE_MOD_PATH, mRootPath)
        ZHTools.swapFragmentWithAnim(
            this,
            SearchModFragment::class.java,
            SearchModFragment.TAG,
            bundle
        )
    }

    private fun parseBundle() {
        val bundle = arguments ?: return
        mRootPath = bundle.getString(BUNDLE_ROOT_PATH, mRootPath)
    }

    private fun initViews() {
        binding.apply {
            mSearchViewWrapper.apply {
                setSearchListener(object : SearchViewWrapper.SearchListener {
                    override fun onSearch(string: String?, caseSensitive: Boolean): Int {
                        return fileRecyclerView.searchFiles(string, caseSensitive)
                    }
                })
                setShowSearchResultsListener(object : SearchViewWrapper.ShowSearchResultsListener {
                    override fun onSearch(show: Boolean) {
                        fileRecyclerView.setShowSearchResultsOnly(show)
                    }
                })
            }

            fileRecyclerView.setFileIcon(FileIcon.MOD)

            operateView.apply {
                addFileButton.setContentDescription(getString(R.string.profile_mods_add_mod))
                createFolderButton.setContentDescription(getString(R.string.profile_mods_download_mod))
                createFolderButton.setImageDrawable(
                    ContextCompat.getDrawable(
                        requireContext(),
                        R.drawable.ic_download
                    )
                )
                pasteButton.setVisibility(if (PasteFile.getInstance().pasteType != null) View.VISIBLE else View.GONE)

                ZHTools.setTooltipText(
                    returnButton,
                    addFileButton,
                    pasteButton,
                    createFolderButton,
                    searchButton,
                    refreshButton
                )
            }
        }
    }

    override fun slideIn(animPlayer: AnimPlayer) {
        binding.apply {
            animPlayer.apply(AnimPlayer.Entry(modsLayout, Animations.BounceInDown))
                .apply(AnimPlayer.Entry(operateLayout, Animations.BounceInLeft))
                .apply(AnimPlayer.Entry(operateButtonsLayout, Animations.FadeInLeft))
        }
    }

    override fun slideOut(animPlayer: AnimPlayer) {
        binding.apply {
            animPlayer.apply(AnimPlayer.Entry(modsLayout, Animations.FadeOutUp))
                .apply(AnimPlayer.Entry(operateLayout, Animations.FadeOutRight))
                .apply(AnimPlayer.Entry(operateButtonsLayout, Animations.BounceShrink))
        }
    }
}

